packages/dd-trace/src/encode/agentless-ci-visibility.js

```diff
@@ -0,0 +1,245 @@
+const Chunk = require('./chunk')
+const log = require('../log')
+const { AgentEncoder } = require('./0.4')
+
+const ENCODING_VERSION = 1
+
+function formatSpan (span) {
+  return {
+    type: span.type === 'test' ? 'test' : 'span',
+    version: ENCODING_VERSION,
+    content: normalizeSpan(truncateSpan(span))
+  }
+}
+
+class AgentlessCiVisibilityEncoder extends AgentEncoder {
+  constructor ({ runtimeId, service, env }) {
+    super(...arguments)
+    this._events = []
+    this.runtimeId = runtimeId
+    this.service = service
+    this.env = env
+    this._traceBytes = new Chunk()
+    this._stringBytes = new Chunk()
+    this._stringCount = 0
+    this._stringMap = {}
+
+    this.reset()
+  }
+
+  count () {
+    return this._events.length
+  }
+
+  append (trace) {
+    this._events = this._events.concat(trace)
+  }
+
+  _encodeValue (bytes, value) {
+    switch (typeof value) {
+      case 'string':
+        this._encodeString(bytes, value)
+        break
+      case 'number':
+        this._encodeNumber(bytes, value)
+        break
+      case 'object':
+        if (Array.isArray(value)) {
+          this._encodeArrayPrefix(bytes, value)
+          for (const event of value) {
+            this._encodeMap(bytes, event)
+          }
+          return
+        }
+        if (value.constructor && value.constructor.name === 'Identifier') {
+          this._encodeId(bytes, value)
+          return
+        }
+        this._encodeMap(bytes, value)
+        break
+      default:
+        // should not happen
+    }
+  }
+
+  _encodeNumber (bytes, value) {
+    if (Math.floor(value) !== value) { // float 64
+      return this._encodeFloat(bytes, value)
+    }
+    return this._encodeInteger(bytes, value)
+  }
+
+  _encodeInteger (bytes, value) {
+    if (value >= 0) {
+      this._encodePositiveLong(bytes, value)
+    } else {
+      this._encodeNegativeLong(bytes, value)
+    }
+  }
+
+  _encodeNegativeLong (bytes, value) {
+    const buffer = bytes.buffer
+    const offset = bytes.length
+
+    if (value >= -0x20) { // negative fixnum
+      bytes.reserve(1)
+      bytes.length += 1
+      buffer[offset] = value
+      return
+    }
+
+    if (value >= -0x80) { // int 8
+      bytes.reserve(2)
+      bytes.length += 2
+      buffer[offset] = 0xd0
+      buffer[offset + 1] = value
+      return
+    }
+
+    if (value >= -0x8000) { // int 16
+      bytes.reserve(3)
+      bytes.length += 3
+      buffer[offset] = 0xd1
+      buffer[offset + 1] = value >> 8
+      buffer[offset + 2] = value
+      return
+    }
+
+    if (value >= -0x80000000) { // int 32
+      bytes.reserve(5)
+      bytes.length += 5
+      buffer[offset] = 0xd2
+      buffer[offset + 1] = value >> 24
+      buffer[offset + 2] = value >> 16
+      buffer[offset + 3] = value >> 8
+      buffer[offset + 4] = value
+      return
+    }
+
+    // int 64
+    const hi = Math.floor(value / Math.pow(2, 32))
+    const lo = value >>> 0
+
+    bytes.reserve(9)
```

Comment:
we might be OK using the biggest possible size only

Reply:
Yes, let's do that. The cost is not much higher, and usually when this method is used it would always be a fairly large number.